firedrake/slate/slac/compiler.py

```diff
@@ -187,10 +187,13 @@ def generate_loopy_kernel(slate_expr, compiler_parameters=None):
                                                        events=events+(slate_loopy_event,))
 
     # map the coefficients in the order that PyOP2 needs
-    new_coeffs = slate_expr.coefficients()
     orig_coeffs = orig_expr.coefficients()
-    get_index = lambda n: orig_coeffs.index(new_coeffs[n]) if new_coeffs[n] in orig_coeffs else n
-    coeff_map = tuple((get_index(n), split_map) for (n, split_map) in slate_expr.coeff_map)
+    new_coeffs = slate_expr.coefficients()
+    map_new_to_orig = [orig_coeffs.index(c) for c in new_coeffs]
+    coeff_map = tuple((map_new_to_orig[n], split_map) for (n, split_map) in slate_expr.coeff_map)
```

Comment:
The index will be the same for `slate_expr.coefficients` and `slate_expr.slate_coefficients`, the only thing that changes it the object with corresponds to the `BlockAssembledVector`, so this should not be a problem. This is relevant for the other comments as well (besides for then one on the coefficients_map)

Reply:
Basically I am using coefficients when I want the original function of the `BlockAssembledVector` and slate_coefficients if I want a `BlockFunction` because I need more information.